### PR TITLE
Permissions grid

### PR DESCRIPTION
**Pull request by sethalves**
- rework domain-level permissions

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4cad94216c7ec1944e815f7572e3bb6ffed5826b available at: https://gist.github.com/319e798badf6e7781fc9a49477cb3c35<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.dmg

**Comment by RyanDowne** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;AIXQUfIy&#x2F;1-add-permissions-grid"> Add permissions grid</a>

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2977c4eab76c38ab2febb1982d357a7119383ebc available at: https://gist.github.com/319e798badf6e7781fc9a49477cb3c35<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: adf893ea1934e1376891b5f7c59127bdc0de86ea available at: https://gist.github.com/319e798badf6e7781fc9a49477cb3c35<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 80eeff5a64f80170e363143daf223b771907fb2c available at: https://gist.github.com/319e798badf6e7781fc9a49477cb3c35<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: be403865fc207f67942da8a3d54ab54a0acdd067 available at: https://gist.github.com/319e798badf6e7781fc9a49477cb3c35<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: dd009bf8b9d08ad31a2189dd03c16ece05dc41de available at: https://gist.github.com/319e798badf6e7781fc9a49477cb3c35<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.dmg

**Comment by ZappoMan** [comment]
One observation I would make is that I previously had a single account listed "satan" and when I started this UI, the order of the accounts was: satan, anonymous, logged-in, localhost... then I added a line item for ZappoMan... the new ordering is: satan, anonymous, logged-in, localhost, zappoman. that ordering seems odd....

I think the "fixed" account lines: localhost, logged-in, and anonymous should be at the top and in that order. all added accounts should follow

**Comment by hifi-gustavo** [comment]
Build console output for commit: a4e4093a18c058d14c4b893653514e9d636195c7 available at: https://gist.github.com/319e798badf6e7781fc9a49477cb3c35<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d202a2bf1170dfd378401a04e1ee84203e13e3e7 available at: https://gist.github.com/319e798badf6e7781fc9a49477cb3c35<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b2f63a2132fc0767fce87738cf62e475132da60b available at: https://gist.github.com/319e798badf6e7781fc9a49477cb3c35<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7696ead0126c9232a0e388c61fb13fe72352c8c4 available at: https://gist.github.com/319e798badf6e7781fc9a49477cb3c35<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.dmg

**Comment by sethalves** [comment]
to test:
- experiment with the permissions settings on http://localhost:40100/settings/
- try logging-in/editing/rezzing/locking with different accounts or with a not-logged-in interface
- try banning an already connected account by giving them a row which doesn't have connection permission.
- etc... lots more.  I'll add soon

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1977efae4f1157ecc8a7bf395780a3dc206788b6 available at: https://gist.github.com/319e798badf6e7781fc9a49477cb3c35<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.dmg

**Comment by kevinmthomas-carpool** [comment]
testing

**Comment by kevinmthomas-carpool** [comment]
## QA Test
**QA Tester**: KevinMThomas
**QA Pass**: **QA Pass** Failure
**Additional Test Results + Observations**: Manipulated ```http://localhost:40100/settings``` by first logging in as anonymous and then by logging in under my account.  Both actions I have manipulated the various check boxes and all results allowed me access to edit the rezzing/locking.  I added my account to the group below and unchecked all boxes however was still able to have full access to rezz and lock.  I reset server and logged out as well to ensure server settings were taken.

**Comment by sethalves** [comment]
@kevinmthomas-carpool did you deny "localhost" the rights before you did this?

File: domain-server/src/DomainServerSettingsManager.cpp [hunk 1]
```diff
@@ -26,6 +28,9 @@
 
 #include "DomainServerSettingsManager.h"
 
+#define WANT_DEBUG 1
```
Comment (ZappoMan) [inline domain-server/src/DomainServerSettingsManager.cpp]: did you mean to leave this in??

**Comment by ZappoMan** [comment]
Code review looks pretty reasonable to me... but maybe @birarda should review it too...

File: domain-server/src/DomainServerSettingsManager.cpp (hunk 1 above)
Comment (sethalves) [inline domain-server/src/DomainServerSettingsManager.cpp]: for now, yes.

**Comment by OmegaHeron** [comment]
Just some notes on this - I'm still working through full testing, but.

Please consider not all of us run from the context of localhost is everything.

I, and others, use Linux (and windows and Mac OS) machines located in remote data centers.  Eventually there needs to be a "localhost equivalent" as in -- ip 1.2.3.4 is treated same as a client trying to connect to a domain via locahost.  That would add same ability to always access a domain regardless of it being at capacity etc etc etc.

**Comment by OmegaHeron** [comment]
Tested Linus - fully remote box located 800+ miles away.  Using assigned users perms works as expected, logged-in users works as expected as does anonymous.

**Comment by hifi-gustavo** [comment]
Build console output for commit: bade215907c4f01e5334e8c885cd245c94f43660 available at: https://gist.github.com/319e798badf6e7781fc9a49477cb3c35<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.dmg

**Comment by OmegaHeron** [comment]
Permissions set to require logged in user presents Incompatible server version vs something helpful like requires you to have HF Identity and be logged in.  I'm sure that will come as this is fully developed.

**Comment by sethalves** [comment]
Are you sure?  It sounds like you've mixed versions.

**Comment by OmegaHeron** [comment]
Wasn't a version mismatch - was a stupid mistake on my end.  When I nuked settings I was an idiot and didn't notice it took me to playa instead of place I wanted to test.  It does indeed say I lack perms to enter.  I'll crawl back under my rock now.

**Comment by OmegaHeron** [comment]
Tested each permutation of options using Interface on Windows 10, stack on Linux both on LAN and over iNet.  All works as expected.  Tested Interface+Stack on Windows over localhost path, all works as expected.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 441b5a20d806206c686ddcfe6b0fed79e21eb8a9 available at: https://gist.github.com/319e798badf6e7781fc9a49477cb3c35<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d54247da8d2f8a28cfd05b6a8b30f70e4a1e186c available at: https://gist.github.com/319e798badf6e7781fc9a49477cb3c35<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.dmg

**Comment by birarda** [comment]
Did a quick pass on this. Looks good to me though I'd think that NodePermissions should be on the wire as a bitfield to save space since we're sending it around a bunch.

**Comment by RyanDowne** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;628TAC1K&#x2F;2-add-lifetime-field-to-domain-settings"> Add lifetime field to domain settings</a>

**Comment by sethalves** [comment]
@birarda it looks like I should have used http://doc.qt.io/qt-4.8/qflags.html for the NodePermissions class.  I have more work in this area queued up, so I'll try to fix that in a future PR.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4c45d081cde51782c27711437ae3670fcf367d39 available at: https://gist.github.com/319e798badf6e7781fc9a49477cb3c35<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8009.dmg

**Comment by OmegaHeron** [comment]
@sethalves - something has gone screwy with this now that it's merged into mainline.
Given the following setup;
DS/AC running on a **second** machine on LAN or DS/AC running on a **second** machine accessed via iNet.

**All anonymous perms off**
**All localhost perms off**
**All logged-in perms off**

**All perms enabled for user OmegaHeron**

I can, as me, enter domain, do various things.  If I login as different user or anon I can not enter domain. Which is correct.  Problem is asset-browser is disabled.  I'm sure this worked before....

**Comment by OmegaHeron** [comment]
Log snip from DS indicates I should be able to see asset-store within Interface (I think)

`[06/11 19:41:39] [DEBUG] Activating local socket for network peer with ID "4f7750ae-8f76-442c-a055-f1ab02d5a5c5"`
`[06/11 19:41:39] [DEBUG] Requesting public key for user "OmegaHeron"`
`[06/11 19:41:39] [DEBUG] Refusing connection from node at 192.168.0.6:56425`
`[06/11 19:41:39] [DEBUG] Username signature matches for "OmegaHeron"`
`[06/11 19:41:39] [DEBUG] user-permissions: specific user matches, so: [permissions: "OmegaHeron" -- connect locks rez rez-tmp asset-server ignore-max-cap]`
`[06/11 19:41:39] [DEBUG] user-permissions: final: [permissions: "OmegaHeron" -- connect locks rez rez-tmp asset-server ignore-max-cap]`
`[06/11 19:41:39] [DEBUG] Added "Agent" (I) {4f7750ae-8f76-442c-a055-f1ab02d5a5c5} x.x.x.x:56425 / 192.168.0.6:56425`

**Comment by OmegaHeron** [comment]
Now I get it - if you drag/drop a file to Interface in a domain where you have proper perms it 1) allows upload 2) opens asset browser.  The fail is in menu item enabling view of asset browser dialog.  Neither its mouse selector or keyboard shortcut work under new code.

**Comment by sethalves** [comment]
okay, I'll take a look.

**Comment by sethalves** [comment]
https://github.com/highfidelity/hifi/pull/8052

**Pull request merged**